Objects/object.c

```diff
@@ -2107,6 +2107,18 @@ _PyTrash_thread_destroy_chain(void)
          * up distorting allocation statistics.
          */
         assert(op->ob_refcnt == 0);
+        /* We need to increase trash_delete_nesting here, otherwise,
+           _PyTrash_thread_destroy_chain will be called recursively
+           and then possibly crash.  An example that may crash without
+           increase:
+               N = 500000  # need to be large enough
+               ob = object()
+               tups = [(ob,) for i in range(N)]
+               for i in range(49):
+                   tups = [(tup,) for tup in tups]
+               del tups
+        */
+        assert(tstate->trash_delete_nesting == 0);
         ++tstate->trash_delete_nesting;
```

Comment:
I think put assert inside could help make sure every chain starting from the deallocators called here have a balanced increase and decrease of trash_delete_nesting. While put outside could only ensure the first chain.